scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -391,6 +439,80 @@ object OptimizationRules {
         WithOnComplete(pipe, ComposedOnComplete(fn0, fn1))
     }
   }
+  /**
+   * a.map(f).flatMap(g) == a.flatMap { x => g(f(x)) }
+   * a.flatMap(f).map(g) == a.flatMap { x => f(x).map(g) }
+   *
+   * This is a rule you may want to apply after having
+   * composed all the maps first
+   */
+  object ComposeMapFlatMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case FlatMapped(Mapped(in, f), g) =>
+        FlatMapped(in, FlatMappedFn(g).runAfter(FlatMapping.Map(f)))
+      case FlatMapValues(MapValues(in, f), g) =>
+        FlatMapValues(in, FlatMappedFn(g).runAfter(FlatMapping.Map(f)))
+      case Mapped(FlatMapped(in, f), g) =>
+        FlatMapped(in, FlatMappedFn(f).combine(FlatMappedFn.fromMap(g)))
+      case MapValues(FlatMapValues(in, f), g) =>
+        FlatMapValues(in, FlatMappedFn(f).combine(FlatMappedFn.fromMap(g)))
+    }
+  }
+
+
+  /**
+   * a.filter(f).flatMap(g) == a.flatMap { x => if (f(x)) g(x) else Iterator.empty }
+   * a.flatMap(f).filter(g) == a.flatMap { x => f(x).filter(g) }
+   *
+   * This is a rule you may want to apply after having
+   * composed all the filters first
+   */
+  object ComposeFilterFlatMap extends Rule[TypedPipe] {
+    def apply[T](on: Dag[TypedPipe]) = {
+      case FlatMapped(Filter(in, f), g) =>
+        Some(FlatMapped(in, FlatMappedFn(g).runAfter(FlatMapping.filter(f))))
+      case filter: Filter[b] =>
+        filter.input match {
+          case fm: FlatMapped[a, b] =>
+            Some(FlatMapped[a, b](fm.input, FlatMappedFn(fm.fn).combine(FlatMappedFn.fromFilter(filter.fn))))
+          case _ => None
+        }
+      case _ =>
+        None
+    }
+  }
+  /**
+   * a.filter(f).map(g) == a.flatMap { x => if (f(x)) Iterator.single(g(x)) else Iterator.empty }
+   * a.map(f).filter(g) == a.flatMap { x => val y = f(x); if (g(y)) Iterator.single(y) else Iterator.empty }
+   *
+   * This is a rule you may want to apply after having
+   * composed all the filters first
+   */
+  object ComposeFilterMap extends Rule[TypedPipe] {
+    def apply[T](on: Dag[TypedPipe]) = {
+      case Mapped(Filter(in, f), g) =>
+        Some(FlatMapped(in, FlatMappedFn.fromFilter(f).combine(FlatMappedFn.fromMap(g))))
+      case filter: Filter[b] =>
+        filter.input match {
+          case fm: Mapped[a, b] =>
+            Some(FlatMapped[a, b](fm.input, FlatMappedFn.fromMap(fm.fn).combine(FlatMappedFn.fromFilter(filter.fn))))
+          case _ => None
+        }
+      case _ =>
+        None
+    }
+  }
+
+  /**
+   * In scalding 0.17 and earlier, descriptions were automatically pushdown below
```

Comment:
For this one, i think we need to do this to enable function coalescing in Dagon/scalding right?

Reply:
Yeah. That’s why I added it. Actually, traps seem to require function coalescing due to the strange way they work. Traps are not totally safe currently it seems and can still throw runtime exceptions if the user confuses cascading.

I think we can make them safe with optimization rules, but they are still confusing since the trap only reaches down to the next write barrier which is not always obvious in the graph. I’m not really a fan of traps by the way.

Reply:
Ha i remember you not being a huge fan of them when they were added. I believe we could implement them in scalding land instead at some point -- and actually just make them work on spark too in the same way ish? some sort of withSideEffectingTry() which is a mix of a flatmap and a try + catch. Dump into a file the bad records. But i looked through the comments related to traps in this pr and it doesn't seem like we have particularly great semantics now and its likely buggy. So thats all just sad. But out of scope for this change to make traps great again.